src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,218 @@
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.previous = previous;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (createHeadOrTailIfNeed(value)) {
+            return;
+        }
+        updateTail(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, true);
+        if (createHeadOrTailIfNeed(value)) {
+            return;
+        }
+        insertNodeAtIndex(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T valueOfNode : list) {
+            add(valueOfNode);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, false);
+        Node<T> foundedNode = foundedNode(index);
+        return index == size ? tail.value : foundedNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, false);
+        Node<T> foundedNode = foundedNode(index);
+        T oldValue = foundedNode.value;
+        foundedNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, false);
+        Node<T> foundedNode = foundedNode(index);
+        T oldValue = foundedNode.value;
+        unlink(foundedNode);
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object || (currentNode.value != null
+                    && currentNode.value.equals(object))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = hasNext(currentNode);
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private Node<T> foundedNode(int index) {
+        if (index == size) {
+            return null;
+        }
+        int distanceFromEnd = size - index;
+        Node<T> nodeFromHead = head;
+        Node<T> nodeFromTail = tail;
+        for (int i = 0; i < (distanceFromEnd >= index ? index : distanceFromEnd
+                - TAIL_ITERATION_CORRECTION); i++) {
+            if (distanceFromEnd >= index) {
+                nodeFromHead = hasNext(nodeFromHead);
+            } else {
+                nodeFromTail = hasPrevious(nodeFromTail);
+            }
+        }
+        return distanceFromEnd >= index ? nodeFromHead : nodeFromTail;
+    }
+
+    private Node<T> hasPrevious(Node<T> node) {
+        return node.previous != null ? node.previous : null;
+    }
+
+    private Node<T> hasNext(Node<T> node) {
+        return node.next != null ? node.next : null;
+    }
```

Comment:
also by initializing the tail and head in the constructor you will not need these checks

Reply:
removed these two methods